src/main/haskell/kore/src/Kore/Step/Condition/Evaluator.hs

```diff
@@ -33,14 +32,30 @@ import qualified Kore.Step.OrOfExpandedPattern as OrOfExpandedPattern
 import           Kore.Step.Simplification.Data
                  ( PureMLPatternSimplifier (..),
                  SimplificationProof (SimplificationProof), Simplifier )
+import           Kore.Step.StepperAttributes
+import           Kore.SMT.SMT
+
+convertStepperToSMT 
+    :: MetadataTools level StepperAttributes 
+    -> MetadataTools level SMTAttributes
+convertStepperToSMT tools = 
+    MetadataTools
+    { symAttributes  = convert . symAttributes  tools
+    , sortAttributes = convert . sortAttributes tools
+    , sortTools = sortTools tools
+    }
+    where convert (StepperAttributes _ _ _ _ _ hook) = SMTAttributes hook
 
 {-| 'evaluate' attempts to evaluate a Kore predicate. -}
 evaluate
-    ::  ( MetaOrObject level
+    ::  forall level variable . 
+        ( MetaOrObject level
         , Given (SortTools level)
         , SortedVariable variable
         , Eq (variable level)
+        , Ord (variable level)
         , Show (variable level)
+        , Given (MetadataTools level StepperAttributes)
```

Comment:
I certainly don't prefer it, but the alternative of threading the argument through every function is worse. Once Cofree annotations are implemented we can get rid of every `Given`.

Reply:
Your are also threading the argument right now. However, doing it explicitly will make people aware of this argument being passed in a local way (i.e. by just looking at the current code, without having to look up function signatures). What I mean is that passing the argument explicitly is like writing documentation, but it does not get out of date.

Also, when we need to send a different thing as argument to some of the called functions, we can easily see where it is needed and we can easily replace it with the new version in exactly those places where we want to.

Reply:
`Given` is actually faster than passing the argument, performance-wise.

Reply:
Ok, most likely, now I have one more reason to rant about Haskell being silly (though I'd have to first understand why Given would be faster).

Regardless, we're not optimizing right now. Even if we're doing some optimizations right now (though that's discouraged if it has an impact on code clarity), this seems like the kind of thing we would want to postpone, since the performance difference we may have due to argument passing stuff is unlikely to be significant for a long time, it if will ever be so. I'm not an expert in Haskell optimizations, but if this difference would be significant in our case, then Haskell would be way too silly.

Anyway, what I said at the end of my previous comment still applies.

Reply:
Actually, after talking to Brandon I'm not sure about `Given` being faster in this case, I may have misinterpreted how it works. 

> Also, when we need to send a different thing as argument to some of the called functions, we can easily see where it is needed and we can easily replace it with the new version in exactly those places where we want to.

Except in this case we're just passing around a big global lookup table. This is precisely the problem that `Given`, `ImplicitParams` and the `Reader` monad all solve.

I originally used `Given` in the smart constructors because that's what Brandon used in his matching-logic branch, and also, it saves boilerplate over passing the `MetadataTools` argument to every smart constructor in an expression like `(mkNot a) `mkAnd` (mkEquals b c)`. As in, not having to write `mkAnd tools (mkNot tools a) (mkEquals tools b c)`. Since I already had a `Given (MetadataTools ...)` constraint running around, it seemed natural to use it in SMT as well, though it wasn't strictly necessary. After the deadline we'll probably refactor this into cofree annotations and/or `Reader`.

Reply:
I agree that the `mk*` code is much nicer this way. I agree that this is a global lookup table. I know that people have been using global variables forever, and every now and then were really sorry that they did so because what was once global was no longer global after the project evolved enough. I agree that `Given` and `Reader` are much nicer than the standard global variables one gets in, say, Java and C++, and, with them, one has much better ways of dealing with a global variable becoming non-global when that happens (though, in my opinion, they are not perfect).

I know that Brandon used this in the matching-logic part of the project, but I don't have an opinion on whether it was better than explicit argument passing (I was just learning about this stuff when I reviewed that code, maybe I should re-read it now).

My main problem was with the fact that it's less visible when something gets passed as an argument (which would be also true for the `Reader` monad, especially if we mix it with other monads like `Counter` and `Error(T)`). I tried to highlight some of the cases where this would matter. My main question is if we gain enough to offset this drop in visibility. For the `mk*` code maybe we did. Here, I'm less sure (which also means that I'm aware that my intuition may be wrong and it may be the best thing to do).

However, we have a deadline, so I just wanted to clarify things. I'm not going to bring any new arguments, I'm just going to approve this, please solve the `Given` issue in whatever way makes sense to you, please check if there are any style issues remaining or any unsolved comments and fix those. After that, as far as I'm concerned, you can merge, unless @ttuegel also has comments.